whats-that-typeclass-Foldable/article.md

```diff
@@ -0,0 +1,315 @@
+
+```haskell
+acc = z
+acc = x_n `f` z
+acc = x_(n-1) `f` (x_n `f` z)
+...
+acc = x_1 `f` (x_2 `f` ( ... (x_(n-1) `f` (x_n `f` z)) ... )
+```
+
+This process could be presented as a picture:
+
+![`foldr` visualisation](./foldr.png)
+
+Unlike `foldr`, `foldl` performs left-to-right folding of a list. Therefore, `f` is applied to `z` and `x_1` firstly — `acc = z 'f' x1`, then `f` is applied to a current value of an accumulator and `x_2` — `acc = (z 'f' x_1) 'f' x_2`, and so on. Finally, `foldl` returns `acc = ( ... ((z 'f' x_1) 'f' x_2) 'f' ... x_(n-1)) 'f' x_n`. 
+
+```markdown
+acc = z
+acc = z `f` x_1
+acc = (z `f` x_1) `f` x_2
+...
+acc = ( ... (z `f` x_1) `f` x_2) `f` ...) `f` x_(n-1)) `f` x_n
+```
+
+The corresponding illustration of these operations:
+
+![`foldl` visualisation](./foldl.png)
+
+### Haskell definition
+
+The recursive Haskell definition of `foldr` is the following:
+
+```haskell
+instance Foldable [] where
+    foldr :: (a -> b -> b) -> b -> [a] -> b
+    foldr _ z []     =  z
+    foldr f z (x:xs) =  x `f` foldr f z xs
+```
+
+## `foldr` and `foldl` simple examples
+
+Let's consider a couple of simple examples of applying `foldr` and `foldl` to a list:
+
+```haskell
+-- addition
+ghci> foldr (+) 0 [1, 2, 3] -- 1 + (2 + (3 + 0))
+6
+ghci> foldl (+) 0 [1, 2, 3] -- ((0 + 1) + 2) + 3
+6
+
+-- exponentiation
+ghci> foldr (^) 2 [2, 3] -- 2 ^ (3 ^ 2)
+512
+ghci> foldl (^) 2 [2, 3] -- (2 ^ 2) ^ 3
+64
+```
+
+It doesn't matter to the addition if the folding is right-to-left or left-to-right as it's an associative operator. However, when it's folding with exponentiation operator, the order is significant.
+
+## Generalization of `foldr` and `foldl`
+
+Fortunately, it's possible to fold not the lists only! We can implement the instance of `Foldable` whenever a data type has one type argument, in other words, when its [kind](https://wiki.haskell.org/index.php?title=Kind&oldid=62154) is `* -> *`. For instance, a list has one type argument — the type of the list's elements:
+
+```haskell
+-- kind of a list
+[] :: * -> *
+
+-- expressions of type list
+[1, 2, 3] :: [Int]
+["a", "b"] :: [String]
+```
+
+`Maybe a` data type also has one type argument — the type of presented value:
+
+```haskell
+-- kind of `Maybe a`
+Maybe :: * -> *
+
+-- expressions of type `Maybe a`
+Just 2 :: Maybe Int
+Just "abc" :: Maybe String
+Nothing :: Maybe a -- couldn't say what type argument `a` is
+```
+
+`Int` and `String` have no type arguments, and you can't obviously, fold them:
+
+```haskell
+-- kind of `Int`
+Int :: *
+-- expressions of type `Int`
+1 :: Int
+10000 :: Int
+
+-- kind of `String`
+String :: *
+-- expressions of type `String`
+"" :: String
+"ab" :: String
+```
+
+`Either a b` data type has two type arguments, corresponding to `Left` and `Right` values, so there is also no reasonable way to fold an `Either a b`:
```

Comment:
✔️

Reply:
Now you say it like if there were no `instance Foldable (Either a)` and `instance Foldable (,) a`, but they in fact exist.

I agree that these instances might be unintuitive, but since they are there, we have to acknowledge them and give some intuition to them. 

(One possible way to explain this: say that elements of the first type argument are ignored, and folding will see only elements of the second type argument, so e.g. `Either l r` will be treated similarly to `Maybe r` by `fold`-like functions)

---

You can also mention, that this can cause bugs when used without caution. E.g. one can write something like `mapM_ verify (toList myTuple)` and hope that both elements will be verified (if they are of the same type), but actually only the second one will be checked. 

I once did this mistake, and after that we added a rule to hlint to check that `Foldable` instance for pairs and eithers is never used.